File: daemon/src/conn.rs
//! Transport-agnostic connection logic: the authentication handshake and the
//! credit-based outbound flow-control pump. Kept independent of BLE so it can be
//! unit-tested on any host; the `ble` module wires raw GATT bytes into these.

use std::collections::{HashMap, VecDeque};

use tracing::warn;

use crate::auth;
use crate::protocol::messages::{AuthChallenge, AuthFail, Hello};
use crate::protocol::{fragment, Frame, Message, Opcode, PROTO_VERSION};
use crate::session::OutEvent;

/// Where a produced frame should be written.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Channel {
    /// `CTRL` characteristic (auth, credits).
    Ctrl,
    /// `P2C` characteristic (session output).
    P2c,
}

/// State of the per-connection HMAC handshake.
#[derive(Debug, Clone, PartialEq, Eq)]
enum AuthState {
    /// Waiting for `HELLO`.
    AwaitingHello,
    /// Challenge sent; waiting for `AUTH_RESPONSE`. Holds the nonce.
    AwaitingResponse([u8; auth::NONCE_LEN]),
    Authenticated,
    Failed,
}

/// Drives the auth handshake on the `CTRL` characteristic.
pub struct AuthHandler {
    psk: Vec<u8>,
    state: AuthState,
    seq: u16,
}

impl AuthHandler {
    pub fn new(psk: Vec<u8>) -> Self {
        Self {
            psk,
            state: AuthState::AwaitingHello,
            seq: 0,
        }
    }

    pub fn is_authenticated(&self) -> bool {
        self.state == AuthState::Authenticated
    }

    /// Process one decoded `CTRL` message. Returns `CTRL` frames to send back.
    /// On a successful `AUTH_RESPONSE` the state flips to authenticated.
    pub fn handle(&mut self, msg: &Message) -> Vec<Frame> {
        match msg.opcode {
            Opcode::Hello => self.on_hello(&msg.payload),
            Opcode::AuthResponse => self.on_auth_response(&msg.payload),
            _ => Vec::new(),
        }
    }

    fn on_hello(&mut self, payload: &[u8]) -> Vec<Frame> {
        if !matches!(self.state, AuthState::AwaitingHello) {
            return Vec::new();
        }
        // Validate version if HELLO parses; tolerate missing fields.
        if let Ok(hello) = serde_json::from_slice::<Hello>(payload) {
            if hello.proto != PROTO_VERSION {
                self.state = AuthState::Failed;
                return self.frame(
                    Opcode::AuthFail,
                    &AuthFail {
                        reason: format!("unsupported proto {}", hello.proto),
                    },
                );
            }
        }
        let nonce = auth::random_nonce();
        self.state = AuthState::AwaitingResponse(nonce);
        self.frame(
            Opcode::AuthChallenge,
            &AuthChallenge {
                nonce: auth::encode_b64(&nonce),
            },
        )
    }

    fn on_auth_response(&mut self, payload: &[u8]) -> Vec<Frame> {
        let AuthState::AwaitingResponse(nonce) = self.state else {
            return Vec::new();
        };
        let resp: crate::protocol::messages::AuthResponse = match serde_json::from_slice(payload) {
            Ok(r) => r,
            Err(_) => {
                self.state = AuthState::Failed;
                return self.frame(
                    Opcode::AuthFail,
                    &AuthFail {
                        reason: "malformed response".into(),
                    },
                );
            }
        };
        if auth::verify(&self.psk, &nonce, &resp.mac) {
            self.state = AuthState::Authenticated;
            self.frame_empty(Opcode::AuthOk)
        } else {
            self.state = AuthState::Failed;
            warn!("auth failed: bad MAC");
            self.frame(
                Opcode::AuthFail,
                &AuthFail {
                    reason: "bad mac".into(),
                },
            )
        }
    }

    fn frame<T: serde::Serialize>(&mut self, opcode: Opcode, value: &T) -> Vec<Frame> {
        let payload = serde_json::to_vec(value).expect("auth payloads serialize");
        let f = Frame {
            opcode,
            session_id: 0,
            flags: 0,
            seq: self.next_seq(),
            payload,
        };
        vec![f]
    }

    fn frame_empty(&mut self, opcode: Opcode) -> Vec<Frame> {
        vec![Frame {
            opcode,
            session_id: 0,
            flags: 0,
            seq: self.next_seq(),
            payload: Vec::new(),
        }]
    }

    fn next_seq(&mut self) -> u16 {
        let s = self.seq;
        self.seq = self.seq.wrapping_add(1);
        s
    }
}

/// Credit-based outbound pump for the `P2C` direction.
///
/// `DATA` and `EXEC_RESULT` frames are flow-controlled per session: each frame
/// costs one credit, and a message is only released once enough credits exist
/// for all its fragments (fragments must stay contiguous). Control opcodes
/// (`SESSION_OPENED`, `CLOSE_SESSION`, `ERROR`, …) bypass credits.
///
/// MVP limitation: a single FIFO means a credit-starved session can head-of-line
/// block others. Per-session queues are a future refinement.
pub struct OutboundPump {
    max_payload: usize,
    seq: u16,
    credits: HashMap<u8, u32>,
    pending: VecDeque<OutEvent>,
}

impl OutboundPump {
    pub fn new(mtu: usize) -> Self {
        Self {
            max_payload: Frame::max_payload(mtu).max(1),
            seq: 0,
            credits: HashMap::new(),
            pending: VecDeque::new(),
        }
    }

    fn is_flow_controlled(opcode: Opcode) -> bool {
        matches!(opcode, Opcode::Data | Opcode::ExecResult)
    }

    fn frame_count(&self, payload_len: usize) -> usize {
        if payload_len == 0 {
            1
        } else {
            payload_len.div_ceil(self.max_payload)
        }
    }

    /// Grant `n` additional credits to a session and release anything now ready.
    pub fn grant(&mut self, session_id: u8, n: u16) -> Vec<Frame> {
        *self.credits.entry(session_id).or_insert(0) += n as u32;
        self.drain()
    }

    /// Submit an outbound event; returns frames ready to write immediately.
    pub fn submit(&mut self, event: OutEvent) -> Vec<Frame> {
        self.pending.push_back(event);
        self.drain()
    }

    /// Number of events still waiting (for tests / metrics).
    pub fn pending_len(&self) -> usize {
        self.pending.len()
    }

    fn drain(&mut self) -> Vec<Frame> {
        let mut out = Vec::new();
        while let Some(front) = self.pending.front() {
            if Self::is_flow_controlled(front.opcode) {
                let need = self.frame_count(front.payload.len()) as u32;
                let have = self.credits.get(&front.session_id).copied().unwrap_or(0);
                if have < need {
                    break; // not enough credits yet — wait
                }
                *self.credits.get_mut(&front.session_id).unwrap() -= need;
            }
            let event = self.pending.pop_front().unwrap();
            out.extend(fragment(
                event.opcode,
                event.session_id,
                event.flags,
                &event.payload,
                self.max_payload,
                &mut self.seq,
            ));
        }
        out
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::protocol::messages::AuthResponse;
    use crate::protocol::Reassembler;

    fn decode_one(frames: &[Frame]) -> Message {
        let mut re = Reassembler::new();
        let mut msg = None;
        for f in frames {
            if let Some(m) = re.push(f.clone()) {
                msg = Some(m);
            }
        }
        msg.expect("one complete message")
    }

    #[test]
    fn full_handshake_succeeds() {
        let psk = auth::random_psk();
        let mut h = AuthHandler::new(psk.to_vec());

        let hello = Message {
            opcode: Opcode::Hello,
            session_id: 0,
            flags: 0,
            payload: serde_json::to_vec(&Hello {
                proto: PROTO_VERSION,
                client: "test".into(),
            })
            .unwrap(),
        };
        let challenge_frames = h.handle(&hello);
        let challenge = decode_one(&challenge_frames);
        assert_eq!(challenge.opcode, Opcode::AuthChallenge);
        let ch: AuthChallenge = serde_json::from_slice(&challenge.payload).unwrap();
        let nonce = auth::decode_b64(&ch.nonce).unwrap();

        let mac = auth::encode_b64(&auth::compute_mac(&psk, &nonce));
        let resp = Message {
            opcode: Opcode::AuthResponse,
            session_id: 0,
            flags: 0,
            payload: serde_json::to_vec(&AuthResponse { mac }).unwrap(),
        };
        let ok = decode_one(&h.handle(&resp));
        assert_eq!(ok.opcode, Opcode::AuthOk);
        assert!(h.is_authenticated());
    }

    #[test]
    fn bad_mac_fails_handshake() {
        let mut h = AuthHandler::new(auth::random_psk().to_vec());
        let hello = Message {
            opcode: Opcode::Hello,
            session_id: 0,
            flags: 0,
            payload: b"{\"proto\":1}".to_vec(),
        };
        h.handle(&hello);
        let resp = Message {
            opcode: Opcode::AuthResponse,
            session_id: 0,
            flags: 0,
            payload: serde_json::to_vec(&AuthResponse {
                mac: auth::encode_b64(&[0u8; 32]),
            })
            .unwrap(),
        };
        let fail = decode_one(&h.handle(&resp));
        assert_eq!(fail.opcode, Opcode::AuthFail);
        assert!(!h.is_authenticated());
    }

    #[test]
    fn version_mismatch_rejected() {
        let mut h = AuthHandler::new(auth::random_psk().to_vec());
        let hello = Message {
            opcode: Opcode::Hello,
            session_id: 0,
            flags: 0,
            payload: b"{\"proto\":99}".to_vec(),
        };
        let fail = decode_one(&h.handle(&hello));
        assert_eq!(fail.opcode, Opcode::AuthFail);
    }

    #[test]
    fn control_opcodes_bypass_credits() {
        let mut pump = OutboundPump::new(100);
        // No credits granted, but SESSION_OPENED must still flow.
        let frames = pump.submit(OutEvent {
            session_id: 1,
            opcode: Opcode::SessionOpened,
            flags: 0,
            payload: b"{\"session_id\":1}".to_vec(),
        });
        assert_eq!(frames.len(), 1);
        assert_eq!(frames[0].opcode, Opcode::SessionOpened);
    }

    #[test]
    fn data_waits_for_credits() {
        let mut pump = OutboundPump::new(100);
        let frames = pump.submit(OutEvent {
            session_id: 1,
            opcode: Opcode::Data,
            flags: 0,
            payload: b"hello".to_vec(),
        });
        assert!(frames.is_empty());
        assert_eq!(pump.pending_len(), 1);

        let released = pump.grant(1, 1);
        assert_eq!(released.len(), 1);
        assert_eq!(released[0].opcode, Opcode::Data);
        assert_eq!(pump.pending_len(), 0);
    }

    #[test]
    fn fragmented_data_needs_credits_for_all_fragments() {
        // mtu 16 -> max_payload 8. 20-byte payload -> 3 frames.
        let mut pump = OutboundPump::new(16);
        let payload = vec![7u8; 20];
        assert!(pump.submit(OutEvent {
            session_id: 2,
            opcode: Opcode::Data,
            flags: 0,
            payload,
        })
        .is_empty());

        // 2 credits is not enough for 3 frames.
        assert!(pump.grant(2, 2).is_empty());
        // One more credit releases the whole message at once.
        let frames = pump.grant(2, 1);
        assert_eq!(frames.len(), 3);
        assert!(frames[0].has_more());
        assert!(!frames[2].has_more());
    }
}

File: daemon/murmurd/conn.py
"""Transport-agnostic connection logic: the auth handshake and the credit-based
outbound flow-control pump. Independent of BLE so it is unit-testable; the
``ble`` module wires raw GATT bytes into these."""

from __future__ import annotations

import json
import logging
import math
from collections import deque
from typing import TYPE_CHECKING

from . import auth
from .protocol import Frame, Message, Opcode, PROTO_VERSION, SeqCounter, fragment

if TYPE_CHECKING:
    from .session import OutEvent

log = logging.getLogger("murmurd.conn")


class AuthHandler:
    """Drives the per-connection HMAC handshake on the CTRL characteristic."""

    def __init__(self, psk: bytes) -> None:
        self._psk = psk
        self._nonce: bytes | None = None
        self._authenticated = False
        self._failed = False
        self._seq = SeqCounter()

    @property
    def is_authenticated(self) -> bool:
        return self._authenticated

    def handle(self, msg: Message) -> list[Frame]:
        if msg.opcode == Opcode.HELLO:
            return self._on_hello(msg.payload)
        if msg.opcode == Opcode.AUTH_RESPONSE:
            return self._on_auth_response(msg.payload)
        return []

    def _on_hello(self, payload: bytes) -> list[Frame]:
        if self._authenticated or self._failed or self._nonce is not None:
            return []
        try:
            hello = json.loads(payload or b"{}")
            if int(hello.get("proto", PROTO_VERSION)) != PROTO_VERSION:
                self._failed = True
                return self._frame(
                    Opcode.AUTH_FAIL, {"reason": f"unsupported proto {hello.get('proto')}"}
                )
        except (ValueError, TypeError):
            pass  # tolerate a malformed HELLO; still challenge
        self._nonce = auth.random_nonce()
        return self._frame(Opcode.AUTH_CHALLENGE, {"nonce": auth.encode_b64(self._nonce)})

    def _on_auth_response(self, payload: bytes) -> list[Frame]:
        if self._nonce is None or self._authenticated:
            return []
        try:
            resp = json.loads(payload)
            mac = resp["mac"]
        except (ValueError, TypeError, KeyError):
            self._failed = True
            return self._frame(Opcode.AUTH_FAIL, {"reason": "malformed response"})
        if auth.verify(self._psk, self._nonce, mac):
            self._authenticated = True
            return self._frame(Opcode.AUTH_OK, {})
        self._failed = True
        expected = auth.encode_b64(auth.compute_mac(self._psk, self._nonce))
        log.warning(
            "auth failed: bad MAC (psk_len=%d nonce=%s got_mac=%r expected_mac=%s)",
            len(self._psk),
            auth.encode_b64(self._nonce),
            mac,
            expected,
        )
        return self._frame(Opcode.AUTH_FAIL, {"reason": "bad mac"})

    def _frame(self, opcode: Opcode, value: dict) -> list[Frame]:
        payload = json.dumps(value, separators=(",", ":")).encode() if value else b""
        return [Frame(opcode, 0, 0, self._seq.next(), payload)]


# Opcodes whose frames are flow-controlled (peripheral -> central data).
_FLOW_CONTROLLED = {Opcode.DATA, Opcode.EXEC_RESULT}

# Cap outstanding credits per session. The central grants generously (and tops
# up on a timer to survive lost notifications); this bounds the window so
# over-granting can't balloon and flow control still means something.
MAX_CREDITS = 96


class OutboundPump:
    """Credit-based outbound pump for the P2C direction.

    DATA/EXEC_RESULT frames cost one credit each; a message is only released once
    enough credits exist for all its fragments (fragments stay contiguous).
    Control opcodes bypass credits. Single FIFO: a credit-starved session can
    head-of-line block others (MVP; per-session queues are a future refinement).
    """

    def __init__(self, mtu: int) -> None:
        from .protocol import max_payload

        self._max_payload = max(1, max_payload(mtu))
        self._seq = SeqCounter()
        self._credits: dict[int, int] = {}
        self._pending: deque[OutEvent] = deque()

    @property
    def pending_len(self) -> int:
        return len(self._pending)

    def _frame_count(self, payload_len: int) -> int:
        return 1 if payload_len == 0 else math.ceil(payload_len / self._max_payload)

    def grant(self, session_id: int, n: int) -> list[Frame]:
        self._credits[session_id] = min(MAX_CREDITS, self._credits.get(session_id, 0) + n)
        return self._drain()

    def submit(self, event: "OutEvent") -> list[Frame]:
        self._pending.append(event)
        return self._drain()

    def _drain(self) -> list[Frame]:
        out: list[Frame] = []
        while self._pending:
            front = self._pending[0]
            if front.opcode in _FLOW_CONTROLLED:
                need = self._frame_count(len(front.payload))
                have = self._credits.get(front.session_id, 0)
                if have < need:
                    break
                self._credits[front.session_id] = have - need
            event = self._pending.popleft()
            out.extend(
                fragment(
                    event.opcode,
                    event.session_id,
                    event.flags,
                    event.payload,
                    self._max_payload,
                    self._seq,
                )
            )
        return out
